### PR TITLE
Support parallell teamcity

### PR DESCRIPTION
**Pull request by Neppord**
The teamcity reporter now reports id for the node and its parent so that the tool reading the output don't need to ques anymore through nesting. This means that it should have full support for parallel test runs.

bellow is a example of the tests running in intellij with the new changes

https://user-images.githubusercontent.com/72190/211170814-3ea8f6b1-21db-4ab0-bd46-7284ffcc5b44.mp4

### CODE REVIEW

**Comment by fsoikin** [comment]
Looks like you need to rebase after I merged https://github.com/purescript-spec/purescript-spec/pull/130

**Comment by Neppord** [comment]
> Looks like you need to rebase after I merged #130

Rebased!

Thanks for all the help!

File: src/Test/Spec/Reporter/TeamCity.purs [hunk 1]
```diff
@@ -38,55 +33,104 @@ escape = Regex.replace'
...
-      [ "name" /\ name
-      , "duration" /\ show (trunc millies)
-      ]
-  Event.TestEnd _ name (Failure error) -> do
-    tellLn $ testFailed
-      [ "name" /\ name
-      , "message" /\ show error
-      ]
-    tellLn $ testFinished [ "name" /\ name ]
+    name <- get <#> Map.lookup path <#> Maybe.fromMaybe ""
```
Comment (fsoikin) [inline src/Test/Spec/Reporter/TeamCity.purs:117]: When would the name for the path be missing? I'm assuming that would indicate a malfunction in the test runner itself, right? In that case, is it really correct to report test end for an empty-named test suite?

File: src/Test/Spec/Reporter/TeamCity.purs [hunk 2]
```diff
@@ -38,55 +33,104 @@ escape = Regex.replace'
...
+withMessage :: String -> ServiceMessage () -> WithMessage
+withMessage message { name, nodeId, parentNodeId } =
+  { name, nodeId, parentNodeId, message }
 
-testFailed :: Array (Tuple String String) -> String
-testFailed = teamcity "testFailed"
+idFromPath :: Path -> String
+idFromPath path = path
+  <#> unwrap
+  <#> (\{ index, name } -> show index <> ":" <> fromMaybe "" name)
```
Comment (fsoikin) [inline src/Test/Spec/Reporter/TeamCity.purs:108]: I wonder if name is really necessary here? From what I understand, `index` is global over the whole universe of tests.

File: src/Test/Spec/Reporter/TeamCity.purs (hunk 1 above)
Comment (Neppord) [reply src/Test/Spec/Reporter/TeamCity.purs:117]: This is a great idea!

What would ve the correct response? How do I implement it well? is `fromJust` the way to go?

File: src/Test/Spec/Reporter/TeamCity.purs (hunk 2 above)
Comment (Neppord) [reply src/Test/Spec/Reporter/TeamCity.purs:108]: I added name to make the output easier to debug by humans. But your right its not necessary though the ":" is to differentiate "0:" from "0", since "0" is a special id that is only allowed as a `parentNodeId` for the top level messages.

Comment (fsoikin) [reply src/Test/Spec/Reporter/TeamCity.purs:108]: Yes, human debugging is a good point. I'm just a bit worried that test names may run afoul of TC's rules for IDs.

File: src/Test/Spec/Reporter/TeamCity.purs (hunk 1 above)
Comment (fsoikin) [reply src/Test/Spec/Reporter/TeamCity.purs:117]: Well, I would just skip reporting if there is no started suite with that path:

```
    mName <- get <#> Map.lookup path
    for_ mName \name ->
      tellLn $ testSuiteFinished $ serviceMessage name path
```

That seems most reasonable: after all, if the path isn't in the map, we must have never issued a "suite started" event.

File: src/Test/Spec/Reporter/TeamCity.purs (hunk 2 above)
Comment (Neppord) [reply src/Test/Spec/Reporter/TeamCity.purs:108]: In what way? that they are to long?

Comment (fsoikin) [reply src/Test/Spec/Reporter/TeamCity.purs:108]: That they may contain characters that TC doesn't like in an ID. Sorry, I only had cursory experience with TC, so I don't really know what rules are.

Comment (Neppord) [reply src/Test/Spec/Reporter/TeamCity.purs:108]: The name is escaped according to spec (as far as I understood it).

Comment (fsoikin) [reply src/Test/Spec/Reporter/TeamCity.purs:108]: Oh I see. 
Ok, good enough for me.

**Review by fsoikin — APPROVED** [review]

**Comment by fsoikin** [comment]
Released https://github.com/purescript-spec/purescript-spec/releases/tag/v7.3.0
